File: dotnet/src/MyDotey.SCF.Simple/Source/StringProperty/Cascaded/CascadedConfigurationSourceConfig.cs
using System;
using System.Collections.Generic;

namespace MyDotey.SCF.Source.StringProperty.Cascaded
{
    /**
     * @author koqizhao
     *
     * May 17, 2018
     */
    public class CascadedConfigurationSourceConfig<C> : DefaultConfigurationSourceConfig
        where C : ConfigurationSourceConfig
    {
        private string _keySeparator;
        private List<string> _cascadedFactors;
        private StringPropertyConfigurationSource<C> _source;

        protected CascadedConfigurationSourceConfig()
        {

        }

        public virtual string KeySeparator { get { return _keySeparator; } }

        public virtual List<string> CascadedFactors { get { return _cascadedFactors; } }

        public virtual StringPropertyConfigurationSource<C> Source { get { return _source; } }

        public override object Clone()
        {
            CascadedConfigurationSourceConfig<C> copy = (CascadedConfigurationSourceConfig<C>)base.Clone();
            if (_cascadedFactors != null)
                copy._cascadedFactors = new List<string>(_cascadedFactors);
            return copy;
        }

        public override string ToString()
        {
            return string.Format("{0} {{ name: {1}, keySeparator: {2}, cascadedFactors: [ {3} ], source {4} }}", GetType().Name,
                Name, _keySeparator, _cascadedFactors == null ? null : string.Join(", ", _cascadedFactors), _source);
        }

        public new class Builder : DefaultConfigurationSourceConfig.DefaultAbstractBuilder<Builder, CascadedConfigurationSourceConfig<C>>
        {
            protected override CascadedConfigurationSourceConfig<C> NewConfig()
            {
                return new CascadedConfigurationSourceConfig<C>();
            }

            public virtual Builder SetKeySeparator(string keySeparator)
            {
                Config._keySeparator = keySeparator;
                return this;
            }

            public virtual Builder AddCascadedFactor(string cascadedFactor)
            {
                if (string.IsNullOrWhiteSpace(cascadedFactor))
                    return this;

                cascadedFactor = cascadedFactor.Trim();
                if (Config._cascadedFactors == null)
                    Config._cascadedFactors = new List<string>();
                Config._cascadedFactors.Add(cascadedFactor);

                return this;
            }

            public virtual Builder AddCascadedFactors(List<string> cascadedFactors)
            {
                if (cascadedFactors != null)
                    cascadedFactors.ForEach(f => AddCascadedFactor(f));

                return this;
            }

            public virtual Builder SetSource(StringPropertyConfigurationSource<C> source)
            {
                Config._source = source;

                return this;
            }

            public override CascadedConfigurationSourceConfig<C> Build()
            {
                if (string.IsNullOrWhiteSpace(Config._keySeparator))
                    throw new ArgumentNullException("keySeparator is null or whitespace");
                Config._keySeparator = Config._keySeparator.Trim();

                if (Config._cascadedFactors == null || Config._cascadedFactors.Count == 0)
                    throw new ArgumentNullException("cascadedFactors is null or empty");

                if (Config._source == null)
                    throw new ArgumentNullException("source is null");

                return base.Build();
            }
        }
    }
}

File: java/src/main/java/org/mydotey/scf/source/cascaded/CascadedConfigurationSourceConfig.java
package org.mydotey.scf.source.cascaded;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import org.mydotey.scf.ConfigurationSource;
import org.mydotey.scf.DefaultConfigurationSourceConfig;

/**
 * @author koqizhao
 *
 * May 17, 2018
 */
public class CascadedConfigurationSourceConfig extends DefaultConfigurationSourceConfig {

    private String _keySeparator;
    private List<String> _cascadedFactors;

    private ConfigurationSource _source;

    protected CascadedConfigurationSourceConfig() {

    }

    public String getKeySeparator() {
        return _keySeparator;
    }

    public List<String> getCascadedFactors() {
        return _cascadedFactors;
    }

    public ConfigurationSource getSource() {
        return _source;
    }

    @Override
    public CascadedConfigurationSourceConfig clone() {
        CascadedConfigurationSourceConfig copy = (CascadedConfigurationSourceConfig) super.clone();

        if (_cascadedFactors != null)
            copy._cascadedFactors = Collections.unmodifiableList(new ArrayList<>(_cascadedFactors));

        return copy;
    }

    @Override
    public String toString() {
        return String.format("%s { name: %s, keySeparator: %s, cascadedFactors: %s }", getClass().getSimpleName(),
            getName(), _keySeparator, _cascadedFactors);
    }

    public static class Builder extends
        DefaultConfigurationSourceConfig.DefaultAbstractBuilder<Builder, CascadedConfigurationSourceConfig> {

        @Override
        protected CascadedConfigurationSourceConfig newConfig() {
            return new CascadedConfigurationSourceConfig();
        }

        public Builder setKeySeparator(String keySeparator) {
            getConfig()._keySeparator = keySeparator;
            return this;
        }

        public Builder addCascadedFactor(String cascadedFactor) {
            if (cascadedFactor == null)
                return this;

            cascadedFactor = cascadedFactor.trim();
            if (cascadedFactor.isEmpty())
                return this;

            if (getConfig()._cascadedFactors == null)
                getConfig()._cascadedFactors = new ArrayList<>();
            getConfig()._cascadedFactors.add(cascadedFactor);

            return this;
        }

        public Builder setSource(ConfigurationSource source) {
            getConfig()._source = source;
            if (getConfig().getName() == null)
                setName("cascaded-" + getConfig().getSource().getConfig().getName());
            return this;
        }

        public Builder addCascadedFactors(List<String> cascadedFactors) {
            if (cascadedFactors != null)
                cascadedFactors.forEach(this::addCascadedFactor);

            return this;
        }

        @Override
        public CascadedConfigurationSourceConfig build() {
            Objects.requireNonNull(getConfig()._keySeparator, "keySeparator is null");
            getConfig()._keySeparator = getConfig().getKeySeparator().trim();
            if (getConfig().getKeySeparator().isEmpty())
                throw new IllegalArgumentException("keySeparator is empty");

            Objects.requireNonNull(getConfig()._cascadedFactors, "cascadedFactors is null");
            if (getConfig().getCascadedFactors().isEmpty())
                throw new IllegalArgumentException("cascadedFactors is empty");

            Objects.requireNonNull(getConfig()._source, "source is null");

            return super.build();
        }
    }

}
